### PR TITLE
[ndr] fix info extraction

### PR DESCRIPTION
**Pull request by remitamine**
i can't test it(cannot download extracted videos urls because of geolocation(i get timeout) i can only access the json request) if someone has access to the site and can run the tests and check if they need to be changed(md5,...).

### CODE REVIEW

**Comment by dstftw** [comment]
Current extraction code works just fine. Patched one fails on http://www.ndr.de/fernsehen/Party-Poette-und-Parade,hafengeburtstag988.html:
```
[debug] System config: []
[debug] User config: []
[debug] Command-line args: [u'http://www.ndr.de/fernsehen/Party-Poette-und-Parade,hafengeburtstag988.html', u'-v']
[debug] Encodings: locale cp1251, fs mbcs, out cp866, pref cp1251
[debug] youtube-dl version 2015.09.03
[debug] Git HEAD: 733196d
[debug] Python version 2.6.6 - Windows-2003Server-5.2.3790-SP2
[debug] exe versions: avconv v12_dev0-1719-g9469370, avprobe v12_dev0-1719-g9469370, ffmpeg N-74386-g542562e, ffprobe N-74386-g542562e, rtmpdump 2.4
[debug] Proxy map: {}
[ndr] hafengeburtstag988: Downloading page
[ndr] hafengeburtstag988: Downloading f4m manifest
ERROR: Unable to download f4m manifest: HTTP Error 403: Forbidden (caused by HTTPError()); please report this issue on https://yt-dl.org/bug . Make sure you are
 using the latest version; see  https://yt-dl.org/update  on how to update. Be sure to call youtube-dl with the --verbose flag and include its complete output.
  File "C:\Dev\git\youtube-dl\master\youtube_dl\extractor\common.py", line 326, in _request_webpage
    return self._downloader.urlopen(url_or_request)
  File "C:\Dev\git\youtube-dl\master\youtube_dl\YoutubeDL.py", line 1865, in urlopen
    return self._opener.open(req, timeout=self._socket_timeout)
  File "C:\python\python266\lib\urllib2.py", line 397, in open
    response = meth(req, response)
  File "C:\python\python266\lib\urllib2.py", line 510, in http_response
    'http', request, response, code, msg, hdrs)
  File "C:\python\python266\lib\urllib2.py", line 435, in error
    return self._call_chain(*args)
  File "C:\python\python266\lib\urllib2.py", line 369, in _call_chain
    result = func(*args)
  File "C:\python\python266\lib\urllib2.py", line 518, in http_error_default
    raise HTTPError(req.get_full_url(), code, msg, hdrs, fp)
```

**Comment by remitamine** [comment]
i change the extractor for two reason:
- using the json request is better than downloading the webpage()
- NJoy is not working with the previous code.

**Comment by dstftw** [comment]
It does not always work, e.g. http://www.ndr.de/nachrichten/niedersachsen/Weil-verspricht-Kommunen-mehr-Geld-fuer-Fluechtlinge,fluechtlinge2944.html, JSON URL http://www.ndr.de/fluechtlinge2944-ppjson.json results in 404.

**Comment by remitamine** [comment]
this page contain an embed for another video it can be detected.
the real video page is:
http://www.ndr.de/fernsehen/sendungen/ndr_aktuell/Norddeutschland-begruesst-Fluechtlinge,ndraktuell28488.html
and this the embed link:
http://www.ndr.de/fernsehen/sendungen/ndr_aktuell/ndraktuell28488-player.html

**Comment by dstftw** [comment]
Therefore webpage has to be fetched in any case. So JSON extraction only make sense as a fallback scenario when webpage extraction fails.

**Comment by remitamine** [comment]
or change NJoy to use the json and keep NDR using the webpage request.

**Comment by dstftw** [comment]
No, n-joy has embeds as well http://www.n-joy.de/entertainment/film/Die-Nominierten-bester-Kurzfilm-Studio-Hamburg,nachwuchspreis102.html.

**Comment by remitamine** [comment]
i gess url doesn't work in both versions.

**Comment by remitamine** [comment]
i can;t parse the info in the web_page the best i can get using js_to_json is:
```
{
1: {"src":"http://hds.ndr.de/z/ndr/2015/0508/TV-20150508-2139-4742.,hi,hq,.mp4.csmil/manifest.f4m", "config": {"OSMFVars": { "plugin_AkamaiAdvancedStreamingPlugin": "http://players.edgesuite.net/flash/plugins/osmf/advanced-streaming-plugin/v3.1/osmf2.0/AkamaiAdvancedStreamingPlugin.swf"}}},
2: {"src":"http://hls.ndr.de/i/ndr/2015/0508/TV-20150508-2139-4742.,lo,hi,hq,.mp4.csmil/master.m3u8", "type":"application/x-mpegURL"},
3: {"src":"http://media.ndr.de/progressive/2015/0508/TV-20150508-2139-4742.hq.mp4", "type":"video/mp4"},
"config": {
"streamType": "httpVideo",
"duration": 3498,
"title": "Party, Pötte und Parade",
"poster": {
1: {"src": "/fernsehen/hafengeburtstag990_v-contentwidescreen.jpg", "quality": "xs"},
2: {"src": "/fernsehen/hafengeburtstag990_v-zweispaltig.jpg", "quality": "s"},
3: {"src": "/fernsehen/hafengeburtstag990_v-contentgross.jpg", "quality": "m"},
4: {"src": "/fernsehen/hafengeburtstag990_v-ardgalerie.jpg", "quality": "l"},
5: {"src": "/fernsehen/hafengeburtstag990_v-contentxl.jpg" || "/common/resources/images/mediathek/default-video-image.jpg", "quality": "xl"}
}
}
}
```
which can't be parsed as json.
so i'm thinking about this:
try to get the json data using the id in the url if it's not possible than search for embed url and pass it to embed extractor else raise an error that there are no media in the page.
i forget to say there is a third reason for the change:
- the previous code doesn't extract all formats

**Comment by remitamine** [comment]
why _extract_f4m_formats doesn't have a fatal argument like _extract_m3u8_formats

**Comment by remitamine** [comment]
i added support for embeds, now it should be able to handle pages that contain ndr or njoy embeds.

**Comment by dstftw** [comment]
`NDRBase` makes no sense since there is no unique `_VALID_URL` can be worked out for it. This causes `TestAllURLsMatching.test_no_duplicates` in `test_all_urls.py` to fail.

**Comment by remitamine** [comment]
> there is no unique _VALID_URL can be worked out for it. 
ndr
```
_VALID_URL = r'https?://www\.ndr\.de//(?:[^/]+/)[A-Z].+,(?P<id>[a-z0-9]+)\.html'
```
ndrembed
```
_VALID_URL = r'https?://www\.n-joy\.de/(?:[^/]+/)(?P<id>[a-z0-9]+)-player'
```
is this enough to be unique
> Because no one have yet added it?
i asked because they are similar in their role, i think it's better to have simulrar arguments.
> Parse it with regex as it's done at the moment
i want to parse all formats like in the json output not only the third format.
i will see if i can get it to work.

**Comment by dstftw** [comment]
For NDREmbed `_VALID_URL` should be similar to:
```
_VALID_URL = r'https?://www\.ndr\.de/(?:[^/]+/)+(?P<id>[a-z0-9]+)-(?:player|externalPlayer)\.html\b'
```
Test examples:
```
http://www.ndr.de/fernsehen/sendungen/nordtour/nordtour7124-player.html
http://www.ndr.de/fernsehen/sendungen/visite/visite11002-player.html
http://www.ndr.de/fernsehen/sendungen/visite/visite11010-externalPlayer.html
http://www.ndr.de/fernsehen/sendungen/hamburg_journal/hamj43006-player.html
http://www.ndr.de/kultur/film/videos/videoimport10424-player.html
http://www.ndr.de/kultur/film/videos/videoimport10422-player.html
http://www.ndr.de/fernsehen/sendungen/schleswig-holstein_magazin/shmag35434-player.html
http://www.ndr.de/fernsehen/sendungen/kulturjournal/kulturjournal4048-player.html
http://www.ndr.de/fernsehen/sendungen/hallo_niedersachsen/hallonds28462-player.html
http://www.ndr.de/info/audio255458-player.html
http://www.ndr.de/fernsehen/roterosen4324-player.html
http://www.ndr.de/fernsehen/livestream/livestream217-externalPlayer.html
http://www.ndr.de/fernsehen/sendungen/ndr_aktuell/ndraktuell28516-player.html
http://www.ndr.de/fernsehen/typisch1030-player.html
http://www.ndr.de/fernsehen/sendungen/weltbilder/weltbilder4518-player.html
http://www.ndr.de/ndrkultur/audio255020-player.html
http://www.ndr.de/fernsehen/doku952-player.html
```

**Comment by remitamine** [comment]
thank you for the help, i will add it when i find a way to make extract_video_info method work for both webpage and json and add it in a single commit.

**Comment by dstftw** [comment]
Well, if we can definitely delegate all article videos to embed extractor (needs extensive testing) - ok let's do it this way, i.e. only extract from JSON.

**Comment by remitamine** [comment]
i know that this way it works in most of the video but it makes some times more requests (like this http://www.ndr.de/nachrichten/niedersachsen/Weil-verspricht-Kommunen-mehr-Geld-fuer-Fluechtlinge,fluechtlinge2944.html) but i think that if it should be optimezed(minimize requests) it's better to be in another pull request.
> Patched one fails on http://www.ndr.de/fernsehen/Party-Poette-und-Parade,hafengeburtstag988.html:
```
[debug] System config: []
[debug] User config: []
[debug] Command-line args: [u'http://www.ndr.de/fernsehen/Party-Poette-und-Parade,hafengeburtstag988.html', u'-v']
[debug] Encodings: locale cp1251, fs mbcs, out cp866, pref cp1251
[debug] youtube-dl version 2015.09.03
[debug] Git HEAD: 733196d
[debug] Python version 2.6.6 - Windows-2003Server-5.2.3790-SP2
[debug] exe versions: avconv v12_dev0-1719-g9469370, avprobe v12_dev0-1719-g9469370, ffmpeg N-74386-g542562e, ffprobe N-74386-g542562e, rtmpdump 2.4
[debug] Proxy map: {}
[ndr] hafengeburtstag988: Downloading page
[ndr] hafengeburtstag988: Downloading f4m manifest
ERROR: Unable to download f4m manifest: HTTP Error 403: Forbidden (caused by HTTPError()); please report this issue on https://yt-dl.org/bug . Make sure you are
 using the latest version; see  https://yt-dl.org/update  on how to update. Be sure to call youtube-dl with the --verbose flag and include its complete output.
  File "C:\Dev\git\youtube-dl\master\youtube_dl\extractor\common.py", line 326, in _request_webpage
    return self._downloader.urlopen(url_or_request)
  File "C:\Dev\git\youtube-dl\master\youtube_dl\YoutubeDL.py", line 1865, in urlopen
    return self._opener.open(req, timeout=self._socket_timeout)
  File "C:\python\python266\lib\urllib2.py", line 397, in open
    response = meth(req, response)
  File "C:\python\python266\lib\urllib2.py", line 510, in http_response
    'http', request, response, code, msg, hdrs)
  File "C:\python\python266\lib\urllib2.py", line 435, in error
    return self._call_chain(*args)
  File "C:\python\python266\lib\urllib2.py", line 369, in _call_chain
    result = func(*args)
  File "C:\python\python266\lib\urllib2.py", line 518, in http_error_default
    raise HTTPError(req.get_full_url(), code, msg, hdrs, fp)
```
to pass this error _extract_f4m_formats should have a fatal and maybe a default argument.
should i add them in this pull request?

**Comment by dstftw** [comment]
If you will extract all formats from the webpage you'll have the same count of extra requests (f4m, m3u8).

**Comment by remitamine** [comment]
i mean in this page(http://www.ndr.de/nachrichten/niedersachsen/Weil-verspricht-Kommunen-mehr-Geld-fuer-Fluechtlinge,fluechtlinge2944.html) the urls of video formats are inside the webpage(1 request) but using this code you have to get the webpage and extract the embed url and then get the json request of the id in embed url(2 requests)

**Comment by dstftw** [comment]
It's only an overhead one JSON page that will be neglected by f4m (possibly recursive) and m3u8.

**Comment by remitamine** [comment]
>```python
_VALID_URL = r'https?://www\.ndr\.de/(?:[^/]+/)+(?P<id>[a-z0-9]+)-(?:player|externalPlayer)\.html\b'
```

i added the regex

File: youtube_dl/extractor/ndr.py [hunk 1]
```diff
@@ -1,77 +1,91 @@
...
-    qualities,
-    parse_duration,
 )
 
 
-class NDRBaseIE(InfoExtractor):
-    def _real_extract(self, url):
-        mobj = re.match(self._VALID_URL, url)
-        video_id = mobj.group('id')
+preferences = {'xl': 4, 'l': 3, 'm': 2, 's': 1, 'xs': 0,}
```
Comment (dstftw) [inline youtube_dl/extractor/ndr.py]: Why not use `qualities`?

Comment (remitamine) [inline youtube_dl/extractor/ndr.py]: i never use it before i will see how it works

**Comment by remitamine** [comment]
now the extractor use qualities to get the preference

**Pull request merged**